src/time.c

```diff
@@ -0,0 +1,276 @@
+// SPDX-License-Identifier: LGPL-2.1
+/*
+ * time.c - efi_time_t helper functions
+ * Copyright 2020 Peter Jones <pjones@redhat.com>
+ */
+
+#include "efivar.h"
+
+int
+efi_time_to_tm(const efi_time_t * const s, struct tm *d)
+{
+
+	if (!s || !d) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	d->tm_year = s->year - 1900;
+	d->tm_mon = s->month - 1;
+	d->tm_mday = s->day;
+	d->tm_hour = s->hour;
+	d->tm_min = s->minute;
+	/*
+	 * Just ignore EFI's range problem here and pretend we're in UTC
+	 * not UT1.
+	 */
+	d->tm_sec = s->second;
+	d->tm_isdst = (s->daylight & EFI_TIME_IN_DAYLIGHT) ? 1 : 0;
+
+	return 0;
+}
+
+int
+tm_to_efi_time(const struct tm * const s, efi_time_t *d, bool tzadj)
+{
+	if (!s || !d) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	d->pad2 = 0;
+	d->daylight = s->tm_isdst ? EFI_TIME_IN_DAYLIGHT : 0;
+	d->timezone = 0;
+	d->nanosecond = 0;
+	d->pad1 = 0;
+	/*
+	 * Just ignore EFI's range problem here and pretend we're in UTC
+	 * not UT1.
+	 */
+	d->second = s->tm_sec;
```

Comment:
Doesn't pretending in this direction create invalid `efi_time_t`? Would it be better to clamp the seconds value to 59?

Reply:
It creates an efi_time_t that is in violation of the spec when you try to represent the leap second.  Yeah, I guess we should clamp it.